presto-common/src/main/java/com/facebook/presto/common/block/ArrayBlockBuilder.java

```diff
@@ -337,4 +338,39 @@ public String toString()
     {
         return format("ArrayBlockBuilder(%d){positionCount=%d}", hashCode(), getPositionCount());
     }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        if (currentEntryOpened) {
+            return false;
+        }
+        ArrayBlockBuilder other = (ArrayBlockBuilder) obj;
+        return this.positionCount == other.positionCount &&
+                Arrays.equals(this.valueIsNull, other.valueIsNull) &&
+                Objects.equals(this.values, other.values) &&
+                Arrays.equals(this.offsets, other.offsets) &&
+                this.hasNullValue == other.hasNullValue &&
+                this.retainedSizeInBytes == other.retainedSizeInBytes;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        if (currentEntryOpened) {
+            return super.hashCode();
+        }
+        return Objects.hash(positionCount,
+                Arrays.hashCode(valueIsNull),
+                values,
+                Arrays.hashCode(offsets),
+                hasNullValue,
+                retainedSizeInBytes);
```

Comment:
Since this hashcode is going to be super costly, I would say include only few fields and not all fields. In my opinion, positionCount, offsets should suffice.

Reply:
> Since this hashcode is going to be super costly, I would say include only few fields and not all fields. In my opinion, positionCount, offsets should suffice.

That will be too confusing. Like we discussed offline this is used *only* at compile time so cost is not a big deal in the grand scheme of things. I say correctness and intuitive implementation is more important.

Reply:
I am a bit conservative here for correctness, as oversimplification may lead to silent optimization failures (in case of possible false positive). Since the hash and equal functions are only used in compile time, maybe we can go with it?

Reply:
@arunthirupathi  How do you think?

Reply:
If 2 different initial entry counts are used, the same content will produce incorrect equals and incorrect hash code. For the compile use cases, if it does not matter, then we can merge this in.